caikit_nlp/toolkit/text_generation/model_run_utils.py

```diff
@@ -0,0 +1,241 @@
+        a certain amount of tokens have been generated. The tuple shall
+        consist of: (start_index, decay_factor) where start_index
+        indicates where penalty starts and decay_factor represents the factor
+        of exponential decay
+    stop_sequences: List[str]:
+        List of strings to be used as stopping criteria
+"""
+
+
+class SequenceStoppingCriteria(StoppingCriteria):
+    def __init__(self, target_sequence_ids):
+        self.target_sequence_ids = target_sequence_ids
+
+    def __call__(self, input_ids, scores, **kwargs):
+        # Check if the target sequence appears in the generated text
+        for seq_id in self.target_sequence_ids:
+            if seq_id in input_ids:
+                return True  # Stop generation
+
+        return False  # Continue generation
+
+    def __len__(self):
+        return 1
+
+    def __iter__(self):
+        yield self
+
+
+def generate_text_func(
+    model,
+    tokenizer,
+    producer_id: ProducerId,
+    eos_token: str,
+    text: str,
+    max_new_tokens: Optional[int] = 20,
+    min_new_tokens: Optional[int] = 0,
+    truncate_input_tokens: Optional[int] = 0,
+    decoding_method: Optional[str] = "GREEDY",
+    top_k: Optional[int] = 0,
+    top_p: Optional[float] = 0.0,
+    typical_p: Optional[float] = 0.0,
+    temperature: Optional[float] = 1.0,
+    repetition_penalty: Optional[float] = 0.0,
+    max_time: Optional[float] = None,
+    exponential_decay_length_penalty: Optional[Tuple[int, float]] = None,
+    stop_sequences: Optional[str] = None,
+    **kwargs
+):
+    __doc__ = """
+        Args:
+            model: PeftModel or transformers.AutoModel
+                Peft model or Transformers model
+            tokenizer: AutoTokenizer
+                Tokenizer to be used with the model
+            producer_id: ProducerId
+                Caikit producer id associated with the module
+            eos_token: str
+                End of sequence token to be used with generation
+            {}
+        Returns:
+            GeneratedTextResult
+    """.format(
+        GENERATE_FUNCTION_ARGS
+    )
+
+    error.type_check("<NLP85452187E>", str, eos_token=eos_token)
+    error.type_check("<NLP65883534E>", str, text=text)
+    error.type_check(
+        "<NLP03860680E>", int, allow_none=True, max_new_tokens=max_new_tokens
+    )
+    error.type_check(
+        "<NLP30091276E>", int, allow_none=True, min_new_tokens=min_new_tokens
+    )
+    error.type_check(
+        "<NLP55411551E>",
+        int,
+        allow_none=True,
+        truncate_input_tokens=truncate_input_tokens,
+    )
+    error.type_check("<NLP84635843E>", int, allow_none=True, top_k=top_k)
+    error.type_check("<NLP55267523E>", float, allow_none=True, top_p=top_p)
+    error.type_check("<NLP13670202E>", float, allow_none=True, typical_p=typical_p)
+    error.type_check(
+        "<NLP11929418E>", float, allow_none=True, repetition_penalty=repetition_penalty
+    )
```

Comment:
Are we leaving generate to do all the validation for values, or would it be better to explicitly check that the max tokens >= min tokens?

Reply:
good point. Will add this validation in here as well